main_loop_tf/main.py

```diff
@@ -417,26 +467,43 @@ def __run(build_model):
                         val_outs,
                         val_summary_ops[s],
                         val_reset_cm_op,
+                        smaller_n_gpus,
                         which_set=s)
 
-
+# Multi-gpu building:
+# In each graph some operation are duplicated so that at runtime
+# it's possible to choose which towers have to be executed. This
+# allows to solve the problem of multi-gpu fault in the case of
+# insufficiently large batches.
+# In this new version the returned operation (for example those for
+# summaries computing) are in the form: [[..], [..]], so that the
+# list of operations for the full and reduced number of gpus used
+# are in the first and second position respectively. In this way
+# it's possible to choose at runtime in main loop or validation
+# which operations of the graph have to be called with respect to
+# the batch size.
+# For what concern the placeholders they now come up as a list of
+# input placeholders associated to each tower that is directly fed
+# from the main loop, so the tf.split is removed and the split
+# is performed out of the graph.
+# Note: the duplicated function are the ones with an 's' as prefix
+# in the name.
+# TODO: double check if this way to operate is correct.
+# In the specific check if duplicating the operation could affect in
+# in some manner the learning task.
 def build_graph(placeholders, input_shape, optimizer, weight_decay, loss_fn,
-                build_model, is_training):
+                build_model, is_training, smaller_n_gpus):
     cfg = gflags.cfg
     devices = cfg.devices
     nclasses = cfg.nclasses
     global_step = cfg.global_step
     if is_training:
-        [inputs, labels, input_split_dim,
-         labels_split_dim, prev_err] = placeholders
+        [inputs_per_gpu, labels_per_gpu, prev_err] = placeholders
     else:
-        [inputs, labels, input_split_dim, labels_split_dim] = placeholders
+        [inputs_per_gpu, labels_per_gpu] = placeholders
 
-    # Split the input among the GPUs (batchwise)
-    inputs_per_gpu = tf.split(inputs, input_split_dim, 0)
-    labels_per_gpu = tf.split(labels, labels_split_dim, 0)
-    for gpu_input in inputs_per_gpu:
-        gpu_input.set_shape(input_shape)
+    labels = tf.concat(labels_per_gpu, axis=0)
+    smaller_labels = tf.concat(labels_per_gpu[:smaller_n_gpus], axis=0)
```

Comment:
Be consistent with the names. If you go for `s_something` use this convention everywhere. `smaller_labels` --> `s_labels`. Also add a comment before the definition of `labels`: concatenate the per-gpu placeholders to get a placeholder for the full (f) list of gpus and one for the subset (s) to be used for the minibatch with less batches.